src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinActivityConversionExtensions.cs

```diff
@@ -184,86 +100,251 @@ private static string EncodeTraceId(ActivityTraceId traceId, bool useShortTraceI
+                    new KeyValuePair<string, object?>(
+                        SpanAttributeConstants.StatusCodeKey,
+                        "ERROR"));
 
-        public void EnumerateTags(Activity activity)
+                // Error flag rule from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk_exporters/zipkin.md#status
+                PooledList<KeyValuePair<string, object?>>.Add(
+                    ref tags,
+                    new KeyValuePair<string, object?>(
+                        ZipkinErrorFlagTagName,
+                        statusDescription));
+            }
+        }
+        else
         {
-            foreach (ref readonly var tag in activity.EnumerateTagObjects())
+            if (GetTag(SpanAttributeConstants.StatusCodeKey) is string status)
             {
-                if (tag.Value == null)
+                if (status == "OK")
                 {
-                    continue;
-                }
-
-                string key = tag.Key;
+                    activity.SetStatus(ActivityStatusCode.Ok);
 
-                if (tag.Value is string strVal)
-                {
-                    PeerServiceResolver.InspectTag(ref this, key, strVal);
-
-                    if (key == SpanAttributeConstants.StatusCodeKey)
-                    {
-                        this.StatusCode = StatusHelper.GetStatusCodeForTagValue(strVal);
-                        continue;
-                    }
-                    else if (key == SpanAttributeConstants.StatusDescriptionKey)
-                    {
-                        // Description is sent as `error` but only if StatusCode is Error. See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk_exporters/zipkin.md#status
-                        this.StatusDescription = strVal;
-                        continue;
-                    }
-                    else if (key == ZipkinErrorFlagTagName)
-                    {
-                        // Ignore `error` tag if it exists, it will be added based on StatusCode + StatusDescription.
-                        continue;
-                    }
+                    PooledList<KeyValuePair<string, object?>>.Add(
+                        ref tags,
+                        new KeyValuePair<string, object?>(
+                            SpanAttributeConstants.StatusCodeKey,
+                            "OK"));
                 }
-                else if (tag.Value is int intVal && tag.Key == SemanticConventions.AttributeNetPeerPort)
+                else if (status == "ERROR")
                 {
-                    PeerServiceResolver.InspectTag(ref this, key, intVal);
+                    string statusDescription =
+                        activity.StatusDescription
+                        ?? GetTag(SpanAttributeConstants.StatusDescriptionKey)
+                        ?? GetTag(ZipkinErrorFlagTagName)
+                        ?? string.Empty;
```

Comment:
Existing code logic set this to `GetTag(SpanAttributeConstants.StatusDescriptionKey) ?? string.Empty` in this code branch (`else` branch of `if (activity.Status != ActivityStatusCode.Unset)`). But I'm not sure if this should change, though.

Reply:
Would `activity.StatusDescription` has value if `activity.Status` is `Unset`?